### PR TITLE
feat: Implement to_json for subset of types

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

Closes https://github.com/apache/datafusion-comet/issues/631

## Rationale for this change

This is part of our effort to support more operations on complex types.

Performance seems ok.

```
AMD Ryzen 9 7950X3D 16-Core Processor
TPCDS Micro Benchmarks:                   Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
to_json                                             147            152           2          1.4         721.3       1.0X
to_json: Comet (Scan)                               141            145           3          1.4         692.0       1.0X
to_json: Comet (Scan, Exec)                          80             88           5          2.6         391.9       1.8X
```

## What changes are included in this PR?

This PR adds an implementation of to_json that works for a subset of types (structs, primitives, strings). There is no support for date or timestamp yet.

## How are these changes tested?

New Rust tests and Spark tests.

### CODE REVIEW

File: docs/source/user-guide/expressions.md [hunk 1]
```diff
@@ -66,32 +66,32 @@ The following Spark expressions are currently available. Any known compatibility
 
 ## String Functions
 
-| Expression      | Notes                                                                                                       |
```
Comment (andygrove) [inline docs/source/user-guide/expressions.md:69]: I ran `prettier` and this caused some formatting changes

**Comment by andygrove** [comment]
@parthchandra could you review?

**Comment by dharanad** [comment]
i would also love to review this. will plan it for tomorrow

**Comment by Kimahriman** [comment]
LGTM

File: native/spark-expr/src/to_json.rs [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+            EvalMode::Legacy,
+            timezone,
+        )?
+        .into_array(arr.len())
+    }
+}
+
+fn struct_to_json(array: &StructArray, timezone: &str) -> Result<ArrayRef> {
+    // get field names
+    let field_names: Vec<String> = array.fields().iter().map(|f| f.name().clone()).collect();
```
Comment (eejbyfeldt) [inline native/spark-expr/src/to_json.rs:126]: This looks like it creates some uneccessary copies of the field names. Any reason to not change this to
```suggestion
    let fields = array.fields();
```
and the usage site then becomes
```
json.push_str(field_names[col_index].name());
```

Or is there some reason to make copies that I am missing?

File: native/spark-expr/src/to_json.rs [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+                    json.push_str(&field_names[col_index]);
+                    json.push_str("\":");
+                    // value
+                    if quotes_needed[col_index] {
+                        json.push('"');
+                    }
+                    json.push_str(string_arrays[col_index].value(row_index));
+                    if quotes_needed[col_index] {
+                        json.push('"');
+                    }
```
Comment (eejbyfeldt) [inline native/spark-expr/src/to_json.rs:179]: I think there is an issue here if the value in `string_arrays[col_index].value(row_index)` contains a literal `"` character. I think spark (or some other json library) would escape such characters.

There are probably also other things like newlines and tabs that also needs to be handled.

File: native/spark-expr/src/to_json.rs [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+                    any_fields_written = true;
+                }
+            }
+            json.push('}');
+            builder.append_value(json.clone());
```
Comment (eejbyfeldt) [inline native/spark-expr/src/to_json.rs:184]: I belive this copy is not needed.
```suggestion
            builder.append_value(&json);
```

File: native/spark-expr/src/to_json.rs [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+            let mut any_fields_written = false;
+            json.push('{');
+            for col_index in 0..string_arrays.len() {
+                if !string_arrays[col_index].is_null(row_index) {
+                    if any_fields_written {
+                        json.push(',');
+                    }
+                    // quoted field name
+                    json.push('"');
+                    json.push_str(&field_names[col_index]);
```
Comment (eejbyfeldt) [inline native/spark-expr/src/to_json.rs:170]: The field_name also needs to be escaped if it contains problematic chars.

**Comment by andygrove** [comment]
Thanks for the review @eejbyfeldt! It is really appreciated. Some very good feedback there. I will address the feedback over the next day or two.

**Comment by parthchandra** [comment]
Sorry @andygrove for this late review. I don't know if one can improve on @eejbyfeldt's review.
To address some of the handling of escape characters, should we look at using something like [serde_json](https://docs.rs/serde_json/latest/serde_json/) ?

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala [hunk 6]
```diff
@@ -1210,6 +1210,58 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde with CometExprShim
...
+                  // TODO implement these types with tests for formatting options and timezone
+                  false
+                case _ => false
+              }
+            }
+
+            val isSupported = child.dataType match {
+              case s: StructType =>
+                s.fields.forall(f => isSupportedType(f.dataType))
+              case _ =>
```
[the remark this answers is not in the document]
Comment (edmondop) [reply spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:1237]: I think @eejbyfeldt point is correct.  Map and Arrays can effectively be jsonified, but they are not `StructType` and we need to extend the pattern matching to support those

**Comment by andygrove** [comment]
@parthchandra @eejbyfeldt This is ready for another review

File: native/spark-expr/src/to_json.rs (hunk 2 above)
Comment (andygrove) [reply native/spark-expr/src/to_json.rs]: This code is now escaping the field names (if needed) so requires a copy in that case

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (andygrove) [reply spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:1293]: I added comments for map/array support

File: native/spark-expr/src/to_json.rs (hunk 3 above)
Comment (andygrove) [reply native/spark-expr/src/to_json.rs]: I have now added escaping for common cases such as \t \r \n \b \f and double quotes

**Review by parthchandra — APPROVED** [review]
lgtm

**Review by huaxingao — APPROVED** [review]
LGTM. Thanks @andygrove

**Pull request merged**